doc/tutorials/basic/caching.md

```diff
@@ -0,0 +1,225 @@
+import panel as pn
+
+from external_module import external_data
+
+pn.extension()
+
+# I'm defined each time the app loads and only shared within that session
+data = {"__name__": __name__}
+
+pn.Column(
+    "## External Module", external_data, f"Object id: {id(external_data)}",
+    "## App", data, f"Object id: {id(data)}",
+).servable()
+```
+
+Run
+
+```bash
+panel serve app.py
+```
+
+Open the app in your browser and refresh it a few times.
+
+It should look like
+
+<video controls="" poster="https://assets.holoviz.org/panel/tutorials/page_load_end.png">
+    <source src="https://assets.holoviz.org/panel/tutorials/page_load.mp4" type="video/mp4" style="max-height: 400px; max-width: 100%;">
+    Your browser does not support the video tag.
+</video>
+
+Please note how the `__name__` and `id` does not change for the `external_module` when you refresh the page. Note that they do change for the served `app`.
+
+:::{note}
+Please note
+
+- imported modules are run once the first time they are imported.
+  - Objects defined in imported modules are shared between all user *sessions*
+- the served `app.py` code is run every time the app is loaded.
+  - Objects defined here are shared within the single user *session* only (unless they are cached).
+- only *specific, bound functions* are rerun on user interactions. Not the entire `app.py` script.
+:::
+
+### Exercise: Add `--autoreload`
+
+Try repeating the steps above with `--autoreload`
+
+```bash
+panel serve app.py --autoreload
+```
+
+What changes?
+
+:::{dropdown} Solution
+
+With `--autoreload` both files are run when the server starts and before the page is loaded the first time.
+
+<img src="https://assets.holoviz.org/panel/tutorials/page_load_end_autoreload.png"></img>
+
+:::
+
+## Load Fast with Caching
+
+Run the code below
+
+```{pyodide}
+from datetime import datetime
+from time import sleep
+
+import panel as pn
+
+pn.extension()
+
+
+def get_data():
+    print("loading data ...")
+    sleep(2)
+    return {"last_update": str(datetime.now())}
+
+
+data = get_data()
+
+pn.pane.JSON(data).servable()
+```
+
+Try refreshing the browser a few times.
+
+Notice that it takes +2 seconds for the application to load and the timestamp is updated every time the app is reloaded.
+
+### Exercise: Apply Caching
+
+Now let's add caching to the `get_data` function by *annotating* it with `@pn.cache`.
+
+Notice in the terminal that the `data` is loaded once when the server starts.
+
+Try loading and refreshing the app several times. It's nice, right?
+
+:::::{dropdown} Solution
+
+::::{tab-set}
+
+:::{tab-item} Annotation
+:sync: annotation
+
+```{pyodide}
+from datetime import datetime
+from time import sleep
+
+import panel as pn
+
+pn.extension()
+
+@pn.cache
+def get_data():
+    print("loading data ...")
+    sleep(2)
+    return {"last_update": str(datetime.now())}
+
+
+data = get_data()
+
+pn.pane.JSON(data).servable()
+```
+
+:::
+
+:::{tab-item} Function
+:sync: function
```

Comment:
I would like to avoid comments where users have to *infer* how to put things together. There is nothing more frustrating in a tutorial than seeing some comment that you cannot figure out how to get working. 

I believe either the example should be there in full. Or it should be removed.

Currently I would like it to be there because I believe its very useful to be able to cache externally given functions without having to wrap them in another function.